extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,234 @@
+				o.logger.Error("OpAMP server returned an error response", zap.String("message", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return o.composeEffectiveConfig(), nil
+			},
+			OnMessageFunc: o.onMessage,
+		},
+		// TODO: Include ReportsEffectiveConfig once the extension has access to the
+		// collector's effective configuration.
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus,
+	}
+
+	if err := o.createAgentDescription(); err != nil {
+		return err
+	}
+
+	if err := o.opampClient.SetAgentDescription(o.agentDescription); err != nil {
+		return err
+	}
+
+	o.logger.Debug("Starting OpAMP client...")
+
+	if err := o.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+
+	o.logger.Debug("OpAMP client started")
+
+	return nil
+}
+
+func (o *opampAgent) Shutdown(ctx context.Context) error {
+	o.logger.Debug("OpAMP agent shutting down...")
+	if o.opampClient == nil {
+		return nil
+	}
+	o.logger.Debug("Stopping OpAMP client...")
+	return o.opampClient.Stop(ctx)
+}
+
+func newOpampAgent(cfg *Config, logger *zap.Logger, build component.BuildInfo, res pcommon.Resource) (*opampAgent, error) {
+	agentType := build.Command
+
+	sn, ok := res.Attributes().Get(semconv.AttributeServiceName)
+	if ok {
+		agentType = sn.AsString()
+	}
+
+	agentVersion := build.Version
+
+	sv, ok := res.Attributes().Get(semconv.AttributeServiceVersion)
+	if ok {
+		agentVersion = sv.AsString()
+	}
+
+	uid := ulid.Make()
+
+	if cfg.InstanceUID != "" {
+		puid, err := ulid.Parse(cfg.InstanceUID)
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	} else {
+		sid, ok := res.Attributes().Get(semconv.AttributeServiceInstanceID)
+		if ok {
+			uuid, err := uuid.Parse(sid.AsString())
+			if err != nil {
+				return nil, err
+			}
+			uid = ulid.ULID(uuid)
+		}
+	}
+
+	agent := &opampAgent{
+		cfg:             cfg,
+		logger:          logger,
+		agentType:       agentType,
+		agentVersion:    agentVersion,
+		instanceID:      uid,
+		effectiveConfig: localConfig, // TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27293
+	}
+
+	return agent, nil
+}
+
+func stringKeyValue(key, value string) *protobufs.KeyValue {
+	return &protobufs.KeyValue{
+		Key: key,
+		Value: &protobufs.AnyValue{
+			Value: &protobufs.AnyValue_StringValue{StringValue: value},
+		},
+	}
+}
+
+func (o *opampAgent) createAgentDescription() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	ident := []*protobufs.KeyValue{
+		stringKeyValue(semconv.AttributeServiceInstanceID, o.instanceID.String()),
+		stringKeyValue(semconv.AttributeServiceName, o.agentType),
+		stringKeyValue(semconv.AttributeServiceVersion, o.agentVersion),
+	}
+
+	nonIdent := []*protobufs.KeyValue{
+		stringKeyValue("os.arch", runtime.GOARCH),
+		stringKeyValue("os.family", runtime.GOOS),
```

Comment:
`os.arch` and `os.family` are not Otel conventions and are not recommended by OpAMP spec. The [spec recommends](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#agentdescriptionnon_identifying_attributes) `os.type` and `os.version` both of which [are Otel conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/os.md).